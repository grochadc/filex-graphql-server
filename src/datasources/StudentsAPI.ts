import { RESTDataSource } from "apollo-datasource-rest";

export type GetStudentFn = (codigo: string, env?: 'dev'|'prod') => Promise<any>;

class StudentsAPI extends RESTDataSource {
  constructor() {
    super();
    this.baseURL = "https://filex-5726c.firebaseio.com/students";
  }

  async _get(url: string, env:string|undefined) {
    const finalUrl = env ? '/'+env+url : '/prod'+url;
    return this.get(finalUrl);
  }

  async getStudent(...[codigo, env]: Parameters<GetStudentFn>) {
    return this.get(`/${env}/${codigo}.json`);
  }
}

export { StudentsAPI };
