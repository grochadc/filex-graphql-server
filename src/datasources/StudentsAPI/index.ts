import { DataSource } from "apollo-datasource";
import { StudentModel } from "../../modules/workshops/models";
import { ParameterizedQuery as PQ } from "pg-promise";
import { ApolloError } from "apollo-server";
import { PrismaClient } from "@prisma/client";
import {
  StudentInput,
  ApplicantInput,
  Applicant,
} from "../../generated/graphql";

type MyStudentInput = Omit<StudentModel, "id">;
type Changes = Partial<MyStudentInput>;

class StudentsAPI extends DataSource {
  db: any;
  prisma: PrismaClient;
  constructor(db: any, prisma: PrismaClient) {
    super();
    if (prisma === undefined)
      throw new Error("Include a new PrismaCLient() on constructor");
    this.prisma = prisma;
    this.db = db;
    if (prisma === undefined) {
      throw Error("Supply a new PrismaCLient() on constructor");
    }
    this.prisma = prisma;
  }

  async getAllStudents(cicloActual: string) {
    const students = await this.prisma.student.findMany({
      where: {
        cicloActual: cicloActual,
        groupId: {
          not: null
        }
      },
      include: {
        applicant: true,
        groupObject: {
          include: {
            teacher: true,
          },
        },
      },
    });

    return students;
  }

  async getStudent(codigo: string, cicloActual: string) {
    if (codigo === null)
      throw new Error("Codigo not provided on studentsAPI.getStudent");
    if (cicloActual === null)
      throw new Error("ciclo_actual not provided on studentsAPI.getStudent");
    const student = await this.prisma.student.findFirst({
      include: {
        applicant: true,
        groupObject: true,
      },
      where: {
        codigo,
        cicloActual,
      },
    });
    if (student === null) throw new Error("Student not found on database");
    if (student.applicant == null)
      throw new Error("Applicant info is missing on database");
    if (student.groupObject == null)
      throw new Error("Couldnt find a group Object");

    return student;
  }

  async addApplicant(
    applicant: ApplicantInput
  ): Promise<Omit<Applicant, "curso" | "groups" | "nivel" | "registering">> {
    const result = await this.prisma.applicant.create({
      data: applicant,
    });

    return {
      ...result,
      id: String(result.id),
    };
  }

  async addStudent(student: StudentInput) {
    const applicant = await this.prisma.applicant.findFirst({
      where: {
        codigo: student.codigo,
      },
    });

    if (applicant === null)
      throw new Error(
        "No applicant found on database. Please submit the student's personal information on the Applicant endpoint first."
      );

    const groupObj = await this.prisma.group.findFirst({
      where: {
        name: student.grupo,
        ciclo: student.cicloActual,
      },
    });
    return Promise.resolve();
    /*
    return this.prisma.student.create({
      data: {
        applicant: { connect: {id: applicant.id} },
        codigo: student.codigo,
        curso: student.curso,
        groupObject: { connect: { id: groupObj.id } },
        ciclo_actual: student.cicloActual,
        nivel: student.nivel
      }
    })
    */
  }

  async editStudent(codigo: string, changes: Changes, ciclo: string) {
    const oldStudent = await this.getStudent(codigo, ciclo);
    const newStudent = Object.assign(oldStudent, changes);
    try {
      this.db.none(
        new PQ({
          text: EDIT_STUDENT,
          values: [
            newStudent.codigo,
            newStudent.nombre,
            newStudent.apellido_materno,
            newStudent.apellido_paterno,
            newStudent.genero,
            newStudent.carrera,
            newStudent.ciclo,
            newStudent.telefono,
            newStudent.email,
            newStudent.nivel,
            newStudent.grupo,
            newStudent.externo,
            newStudent.curso,
            newStudent.id,
          ],
        })
      );
    } catch (e) {
      console.error(e);
    }
    return newStudent;
  }
}

export const GET_STUDENT = `
  SELECT * FROM Student WHERE codigo=$1
`;

export const ADD_STUDENT = `
  INSERT INTO Student (
    codigo,
    nombre,
    apellido_materno,
    apellido_paterno,
    genero,
    carrera,
    ciclo,
    telefono,
    email,
    nivel,
    grupo,
    externo,
    curso
  ) values (
    $1,
    $2,
    $3,
    $4,
    $5,
    $6,
    $7,
    $8,
    $9,
    $10,
    $11,
    $12,
    $13
  )
  RETURNING id
`;

export const EDIT_STUDENT = `
  UPDATE Student SET
  codigo=$1,
  nombre= $2,
  apellido_materno= $3,
  apellido_paterno= $4,
  genero= $5,
  carrera= $6,
  ciclo= $7,
  telefono= $8,
  email= $9,
  nivel= $10,
  grupo= $11,
  externo= $12,
  curso= $13
  WHERE id=$14
`;

export { StudentsAPI };
