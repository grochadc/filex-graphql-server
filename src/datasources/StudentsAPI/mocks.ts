import { ParameterizedQuery as PQ } from "pg-promise";
import { ADD_STUDENT, GET_STUDENT } from "./index";

export const applicants = [
  {
    id: 1,
    codigo: "1234567890",
    nombre: "Benito Antonio",
    apellido_paterno: "Martinez",
    apellido_materno: "Ocasio",
    genero: "M",
    carrera: "Abogado",
    ciclo_ingreso: "2021A",
    telefono: "1234567890",
    email: "bad@bunny.pr",
    externo: false,
  },
  {
    id: 2,
    codigo: "0987654321",
    nombre: "Alberto",
    apellido_paterno: "Aguilera",
    apellido_materno: "Valadez",
    genero: "M",
    carrera: "Abogado",
    ciclo_ingreso: "2021A",
    telefono: "1234567890",
    email: "juanga@elnoanoa.mx",
    externo: false,
  },
];

export const students = [
  {
    id: 1,
    applicantId: 1,
    ciclo_actual: "2022B",
    codigo: "1234567890",
    nivel: 4, 
    curso: "en",
    groupId: 1,
    desertor: false,
    situacion: null,
  },
  {
    id: 2,
    applicantId: 2,
    ciclo_actual: "2022B",
    codigo: "0987654321",
    nivel: 3, 
    curso: "en",
    groupId: 2,
    desertor: false,
    situacion: null,
  }
]

const groups = [
  {
    id: 1,
    ciclo: "2022B",
    name: "E4-1",
    time: "07:00-08:00",
    aula: "G5",
    teacherId: 1,
    nivel: 4,
    registrados: 1,
    course: "en"
  },
  {
    id: 2,
    ciclo: "2022B",
    name: "E3-1",
    time: "07:00-08:00",
    aula: "G2",
    teacherId: 2,
    nivel: 3,
    registrados: 1,
    course: "en"
  }
];


export const studentss = [
  {
    id: 1,
    codigo: "1234567890",
    nombre: "Benito Antonio",
    apellido_paterno: "Martinez",
    apellido_materno: "Ocasio",
    genero: "M",
    carrera: "Abogado",
    ciclo: "2021A",
    telefono: "1234567890",
    email: "bad@bunny.pr",
    nivel: "4",
    curso: "en",
    grupo: "E4-1",
    externo: false,
  },
  {
    id: 2,
    codigo: "0987654321",
    nombre: "Alberto",
    apellido_paterno: "Aguilera",
    apellido_materno: "Valadez",
    nivel: "1",
    grupo: "E1-1",
    genero: "M",
    carrera: "Abogado",
    ciclo: "2021A",
    telefono: "1234567890",
    email: "juanga@elnoanoa.mx",
    curso: "en",
    externo: false,
  },
];

const mocks = {
  db: {
    one: jest.fn((pq: PQ) => {
      if (pq.text === ADD_STUDENT) return Promise.resolve({ id: 1 });
    }),
    none: jest.fn((pq) => {
      return Promise.resolve();
    }),
    oneOrNone: jest.fn((pq) => {
      if (pq.text === GET_STUDENT) {
        const student = students.find(
          (student) => student.codigo === pq.values[0]
        );
        return Promise.resolve(student ? student : null);
      }
    }),
    any: jest.fn((pq) => {
      if (pq.text === GET_STUDENT) {
        const student = students.find(
          (student) => student.codigo === pq.values[0]
        );
        return Promise.resolve(student ? student : null);
      }
    }),
  },
};

export default mocks;
