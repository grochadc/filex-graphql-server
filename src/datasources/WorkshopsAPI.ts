import { RESTDataSource } from "apollo-datasource-rest";
import * as R from "ramda";
import * as utils from "../utils";
import database, { Option } from "./db";
import { Workshop, AttendingStudent } from "../generated/graphql";
import {
  WorkshopModel,
  TeacherModel,
  OptionModel,
  StudentReservationModel,
  TeacherReservationModel,
  DatabaseModel,
} from "../modules/workshops/models";

type Maybe<T> = T | null;

class WorkshopsAPI extends RESTDataSource {
  constructor() {
    super();
    this.baseURL = "https://filex-5726c.firebaseio.com/workshops/";
  }

  async isOpen(): Promise<boolean> {
    const open = await this.get(
      `${this.context.enviroment}/system/isOpen/result.json`
    );
    return open ? open : false;
  }

  async toggleOpen(): Promise<boolean> {
    const open = await this.isOpen();
    await this.put(`${this.context.enviroment}/system/isOpen.json`, {
      result: !open,
    });
    return !open;
  }

  async getWorkshops(): Promise<WorkshopModel[]> {
    const workshops = await this.get(
      `${this.context.enviroment}/workshops.json`
    );
    if (workshops === null)
      throw new Error(
        `${this.context.enviroment}/workshops.json returned null`
      );
    return Object.values(workshops);
  }

  async getOptions(): Promise<DatabaseModel["options"]> {
    return (await this.get(
      `${this.context.enviroment}/options.json`
    )) as DatabaseModel["options"];
  }

  getAvailableOptions(): Promise<Maybe<DatabaseModel["availableOptions"]>> {
    return this.get(
      `${this.context.enviroment}/availableOptions.json`
    ) as Promise<Maybe<DatabaseModel["availableOptions"]>>;
  }

  getTeacher(id: string): Promise<TeacherModel> {
    return this.get(`${this.context.enviroment}/teachers/${id}.json`);
  }

  getStudentReservation(codigo: string): Promise<StudentReservationModel> {
    return this.get(
      `/${this.context.enviroment}/studentsReservations/${codigo}.json`
    );
  }

  async makeReservation(
    codigo: string,
    teacher_id: string,
    option_id: string,
    tutorial_reason?: string
  ) {
    const option = await this.getOptionById(option_id);
    if (option === null)
      throw new Error("option with id " + option_id + " returned null");
    const student = await this.context.dataSources.studentsAPI.getStudent(
      codigo
    );
    const { nombre, apellido_paterno, apellido_materno, nivel, grupo } =
      student;
    const teacherReservation: TeacherReservationModel = {
      codigo: student.codigo,
      nombre,
      apellido_paterno,
      apellido_materno,
      nivel,
      grupo,
      option_id: option.id,
      option_name: "why? option_name from makeReservation dataSource",
      workshop_id: option.workshop_id,
      workshop_name: option.workshop_name,
      tutorial_reason: tutorial_reason ? tutorial_reason : null,
    };
    this.post(
      `${this.context.enviroment}/teachers/${teacher_id}/raw_reservations/${option_id}.json`,
      teacherReservation
    );
    this.put(`${this.context.enviroment}/studentsReservations/${codigo}.json`, {
      option_id: option_id,
    });

    //enqueue
    const availableNumber = await this.get(
      `${this.context.enviroment}/availableOptions/${option_id}.json`
    );
    this.put(
      `${this.context.enviroment}/availableOptions/${option_id}.json`,
      availableNumber + 1
    );

    return { ...option, option_id: option.id };
  }

  deleteReservations(teacher_id: string, option_id: string) {
    return this.delete(
      `${this.context.enviroment}/teachers/${teacher_id}/raw_reservations/${option_id}.json`
    );
  }

  async resetReservations() {
    await this.delete(`${this.context.enviroment}/studentsReservations.json`);
    await this.delete(`${this.context.enviroment}/availableOptions.json`);
    return true;
  }

  getOptionById(id: string): Promise<OptionModel> {
    return this.get(`${this.context.enviroment}/options/${id}.json`);
  }

  async saveAttendance(attendance: AttendingStudent[]) {
    const date = new Date();
    const values = attendance.map((student) => {
      return [
        `=date(${date.getFullYear()},${date.getMonth() + 1},${date.getDate()})`,
        student.codigo,
        student.nombre,
        student.apellido_paterno,
        student.apellido_materno,
        student.nivel,
        student.grupo,
        student.workshop,
        student.teacher,
        student.attended,
      ];
    });
    const range = "Attendance!A1";

    try {
      await this.context.dataSources.workshopsSheetsAPI.append(values, range);
    } catch (e) {
      console.error(e);
    }
    return true;
  }

  setWorkshopLink(option_id: string, url: string) {
    this.put(`${this.context.enviroment}/options/${option_id}/url.json`, url);
    return true;
  }

  async getMaxStudentReservations() {
    const result = await this.get(`${this.context.enviroment}/system.json`);
    return result.max_reservations;
  }

  async getReservationLimit(): Promise<number> {
    return this.get("/reservationLimit.json");
  }
}

export { WorkshopsAPI };
