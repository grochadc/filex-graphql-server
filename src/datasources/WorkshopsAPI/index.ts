import { RESTDataSource } from "apollo-datasource-rest";
import { AttendingStudent } from "../../generated/graphql";
import {
  WorkshopModel,
  OptionModel,
  DatabaseModel,
} from "../../modules/workshops/models";
import { PrismaClient } from "@prisma/client";
import { ApolloError } from "apollo-server";
import { students } from "datasources/StudentsAPI/mocks";

type Maybe<T> = T | null;

class WorkshopsAPI extends RESTDataSource {
  prisma: PrismaClient;
  constructor(prisma: PrismaClient) {
    super();
    if (prisma === undefined)
      throw new Error("Must include a new PrismaClient() on constrctor");
    this.prisma = prisma;
    this.baseURL = "https://filex-5726c.firebaseio.com/workshops/";
    if (process.env.NODE_ENV === "test") this.initialize({} as any);
  }

  getTodaysDate(): Date {
    return new Date();
  }

  async isOpen(): Promise<boolean> {
    const open = await this.get(
      `${this.context.enviroment}/system/isOpen/result.json`
    );
    return open ? open : false;
  }

  async toggleOpen(): Promise<boolean> {
    const open = await this.isOpen();
    await this.put(`${this.context.enviroment}/system/isOpen.json`, {
      result: !open,
    });
    return !open;
  }

  getTeacher(id: string) {
    return this.prisma.teacher.findUnique({
      where: {
        id: Number(id),
      },
    });
  }

  async getAllTeachers() {
    return this.prisma.teacher.findMany({
      include: {
        options: {
          include: {
            workshop: true,
          },
        },
      },
    });
  }

  async getReservationsByOptionId(option_id: string) {
    const openDate = await this.getOpenDate();
    const reservations = await this.prisma.workshopReservation.findMany({
      where: {
        option_id: Number(option_id),
        create_time: {
          gte: new Date(openDate),
        },
      },
      include: {
        student: {
          include: {
            applicant: true,
          },
        },
        option: true,
      },
    });

    if (null) return [];

    return reservations.map((reservation) => {
      return {
        ...reservation,
        id: String(reservation.id),
      };
    });
  }

  async getReservationCount(student_id: string) {
    return this.prisma.workshopReservation.count({
      where: {
        student_id: Number(student_id),
      },
    });
  }

  async getOptionReservationCount(option_id: string) {
    const openDate = await this.getOpenDate();

    return this.prisma.workshopReservation.count({
      where: {
        option_id: Number(option_id),
        create_time: {
          gte: new Date(openDate),
        },
      },
    });
  }

  async getAllOptions(max_students: number) {
    return this.prisma.workshopOption.findMany();
  }

  async getAllWorkshops(max_students: number) {
    return this.prisma.workshop.findMany({
      include: {
        options: {
          include: {
            teacher: true,
          },
        },
      },
    });
  }

  async getTeacherReservations(teacherId: number, optionId: number) {
    const openDateStr = await this.getOpenDate();
    return this.prisma.workshopReservation.findMany({
      where: {
        option: {
          teacher_id: teacherId,
          id: optionId,
        },
        create_time: {
          gt: new Date(openDateStr),
        },
      },
      include: {
        option: true,
        student: {
          include: {
            applicant: true,
            groupObject: true,
          },
        },
      },
    });
  }

  async getTeacherOptions(teacherId: string) {
    return this.prisma.workshopOption.findMany({
      include: {
        workshop: true,
      },
      where: {
        teacher_id: Number(teacherId),
      },
    });
  }

  async getStudentReservation(studentId: string) {
    const openDateString = await this.getOpenDate();
    const reservation = await this.prisma.workshopReservation.findFirst({
      where: {
        student_id: Number(studentId),
        create_time: {
          gte: new Date(openDateString),
        },
      },
      include: {
        student: true,
        option: {
          include: {
            workshop: true,
            teacher: true,
          },
        },
      },
    });
    if (!reservation) return null;
    return { ...reservation, id: String(reservation.id) };
  }

  async makeReservation(student_id: string, option_id: string) {
    if (!student_id)
      throw new Error(
        "No student_id provided to make a reservation on workshopsAPI.makeReservation"
      );
    if (!option_id)
      throw new Error(
        "No option_id provided to make a reservation on workshopsAPI.makeReservation"
      );
    const currentReservation = await this.getStudentReservation(student_id);
    if (currentReservation)
      throw new ApolloError("Student already has a reservation", "409");
    const reservation = await this.prisma.workshopReservation.create({
      data: {
        student_id: Number(student_id),
        option_id: Number(option_id),
      },
      include: {
        student: true,
        option: {
          include: {
            teacher: true,
            workshop: true,
          },
        },
      },
    });
    return { ...reservation, id: String(reservation.id) };
  }

  async resetReservations() {
    const now = await this.getTodaysDate();
    const nowString = now.toISOString();
    await this.put("system/openDate.json", { result: nowString });
    return true;
  }

  async getOpenDate(): Promise<string> {
    const openDate = await this.get("system/openDate.json");
    const dateString = openDate.result;
    return dateString;
  }

  async saveAttendance(attendance: AttendingStudent[]) {
    const reservation_ids = attendance
      .filter((student) => student.attended)
      .map((student) => Number(student.reservation_id));
    const modified = await this.prisma.workshopReservation.updateMany({
      where: {
        id: {
          in: reservation_ids,
        },
      },
      data: {
        attended: true,
      },
    });
    return modified.count > 0;
  }

  async getMaxStudentReservations() {
    const result = await this.get(`system.json`);
    return result.max_reservations;
  }

  async getReservationLimit(): Promise<number> {
    return this.get("/reservationLimit.json");
  }
}

export { WorkshopsAPI };
