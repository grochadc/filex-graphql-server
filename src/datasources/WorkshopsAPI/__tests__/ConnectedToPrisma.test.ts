//import prisma from '../../testutils/prisma/client';
import { WorkshopsAPI } from "../../WorkshopsAPI";
import { StudentsAPI } from "../../StudentsAPI";
import * as dotenv from "dotenv";
dotenv.config({
  path: ".env.test",
});
import * as fs from "fs/promises";
import * as path from "path";
import { PrismaClient } from "@prisma/client";

if (process.env.DOTENV_ENV !== "test") {
  console.log("Init prisma with url", process.env.HEROKU_POSTGRESQL_TEAL_URL);
  throw new Error(
    `Careful! If this script runs with a production db string reservations will be erased at the end!!/nCheck the file .env.test is being read by Jest instead of .env`
  );
}

const GENERATE_MOCKS = false;
const MOCKS_PATH = "../mocks";
function writeMock(filename: string, data: any) {
  fs.writeFile(
    path.join(__dirname, MOCKS_PATH, filename + ".js"),
    `module.exports = ${JSON.stringify(data, null, 2)};`
  )
    .then(() => console.log(`${filename} file written!`))
    .catch((e) => console.error(e));
}
if (GENERATE_MOCKS) console.log("Generating Mocks using Prisma Connection");

const prisma = new PrismaClient({
  datasources: {
    db: {
      url: process.env.HEROKU_POSTGRESQL_TEAL_URL,
    },
  },
});

const workshopsAPI = new WorkshopsAPI(prisma);
const studentsAPI = new StudentsAPI({} as any, prisma);

workshopsAPI.getOpenDate = jest.fn(() => {
  const openDate = new Date(2021, 8, 12);
  return Promise.resolve(openDate.toISOString());
});

test("gets a student with applicant info", async () => {
  const student = await studentsAPI.getStudent("1234567890", "2022B");
  expect(student.ciclo_actual).toBe("2022B");
  expect(student.codigo).toBe("1234567890");
  expect(student).toMatchSnapshot();

  if (GENERATE_MOCKS) {
    writeMock("getStudentPrismaMock", student);
  }
});

test("makes a reservation", async () => {
  workshopsAPI.getTodaysDate = jest.fn(() => new Date(2022, 8, 13));
  const studentId = "3";
  const optionId = "1";

  const reservation = await workshopsAPI.makeReservation(studentId, optionId);
  expect(reservation).toBeDefined();
  expect(reservation.option_id).toBe(1);
  expect(reservation.student.id).toBe(3);
  expect(reservation).toMatchSnapshot();

  if (GENERATE_MOCKS) {
    writeMock("makeReservationPrismaMock", reservation);
  }

  const studentReservation = await prisma.workshopReservation.findFirst({
    where: {
      student_id: Number(studentId),
      create_time: {
        gt: new Date(2000, 1, 1),
      },
    },
  });
  expect(studentReservation).toBeDefined();
  expect(String(studentReservation?.id)).toBe(reservation.id);
});

test("gets a students reservation", async () => {
  await prisma.workshopReservation.create({
    data: {
      student_id: 3,
      option_id: 1,
      attended: false,
    },
  });

  const reservation = await workshopsAPI.getStudentReservation("3");
  expect(reservation).not.toBeNull();
  expect(reservation?.student.codigo).toBe("1234567890");
  expect(reservation?.option.day).toBe("Lunes");
  expect(reservation?.option.workshop.name).toBe("Conversation");
  expect(reservation?.option.teacher.nombre).toBe("Gonzalo Rocha");

  if (GENERATE_MOCKS) {
    writeMock("getStudentReservationPrismaMock", reservation);
  }
});

test("gets a teacher reservation list", async (teacherId = "1") => {
  workshopsAPI.getTodaysDate = jest.fn(() => {
    return new Date(2022, 9, 21);
  });
  await prisma.workshopReservation.create({
    data: {
      student_id: 3,
      option_id: 1,
    },
  });
  const options = await workshopsAPI.getTeacherOptions(teacherId);
  const reservations = await workshopsAPI.getTeacherReservations(
    Number(teacherId),
    options[0].id
  );

  expect(reservations[0].student.applicant.nombre).toBe("Benito Antonio");
  expect(reservations[0].student.id).toBe(3);
  expect(reservations[0].option.id).toBe(options[0].id);

  if (GENERATE_MOCKS) {
    writeMock("getTeacherOptionsPrismaMock", options);
    writeMock("getTeacherReservationsPrismaMock", reservations);
  }
});

test("gets a teacher option with all reservations for attendance", async () => {
  await prisma.workshopReservation.create({
    data: {
      student_id: 2,
      option_id: 1,
    },
  });

  await prisma.workshopReservation.create({
    data: {
      student_id: 3,
      option_id: 1,
    },
  });

  const attendanceList = await workshopsAPI.getTeacherReservations(1, 1);
  expect(attendanceList[0].student.id).toBe(2);
  expect(attendanceList[0].student.applicant.nombre).toBe("Alberto");
  expect(attendanceList[0].student.groupObject?.name).toBe("E4-1");
  expect(attendanceList[0].option.day).toBe("Lunes");
  expect(attendanceList[0].attended).toBe(false);
  expect(attendanceList).toMatchSnapshot();

  if (GENERATE_MOCKS) writeMock("getTeacherReservationsMock", attendanceList);
});

test("api.getTeacher", async () => {
  const teacher = await workshopsAPI.getTeacher("1");
  expect(teacher).toBeDefined();
  expect(teacher?.nombre).toBe("Gonzalo Rocha");

  if (GENERATE_MOCKS) writeMock("getTeacherPrismaMock", teacher);
});

test("get all workshops", async () => {
  const allWorkshops = await workshopsAPI.getAllWorkshops(30);
  expect(allWorkshops).toHaveLength(1);
  expect(allWorkshops[0].options).toHaveLength(1);
  expect(allWorkshops[0].options[0].teacher).toBeDefined();

  if (GENERATE_MOCKS) writeMock("getAllWorkshopsPrismaMock", allWorkshops);
});

test("gets all teachers", async () => {
  const allTeachers = await workshopsAPI.getAllTeachers();
  expect(allTeachers).toHaveLength(1);
  expect(allTeachers[0].nombre).toBe("Gonzalo Rocha");
  expect(allTeachers[0].options).toHaveLength(1);
  expect(allTeachers[0].options[0].workshop.name).toBe("Conversation");
  if (GENERATE_MOCKS) writeMock("getAllTeachersPrismaMock", allTeachers);
});

afterEach(async () => {
  await prisma.workshopReservation.deleteMany({});
  await prisma.$executeRaw`
        ALTER SEQUENCE workshop_reservation_id_seq RESTART WITH 1
    `;
});

afterAll(async () => {
  await prisma.$disconnect();
});
