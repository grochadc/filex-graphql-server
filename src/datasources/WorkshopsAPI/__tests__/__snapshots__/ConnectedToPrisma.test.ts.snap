// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gets a student with applicant info 1`] = `
{
  "applicant": {
    "apellido_materno": "Ocasio",
    "apellido_paterno": "Martinez",
    "carrera": "Abogado (DECH)",
    "ciclo_ingreso": "2022A",
    "codigo": "1234567890",
    "email": "bad@bunny.pr",
    "externo": false,
    "genero": "M",
    "id": 1,
    "institutional_email": "benito.martinez@cusur.udg.mx",
    "nombre": "Benito Antonio",
    "telefono": "0001234567",
  },
  "applicantId": 1,
  "ciclo_actual": "2022B",
  "codigo": "1234567890",
  "curso": "en",
  "desertor": false,
  "groupId": 2,
  "groupObject": {
    "aula": "F4",
    "ciclo": "2022B",
    "course": "en",
    "id": 2,
    "name": "E4-1",
    "nivel": 4,
    "registrados": 30,
    "teacherId": 1,
    "time": "08:00-09:00",
  },
  "id": 3,
  "nivel": 4,
  "situacion": null,
}
`;

exports[`gets a teacher option with all reservations for attendance 1`] = `
{
  "attended": false,
  "create_time": Any<Date>,
  "id": 1,
  "option": {
    "active": true,
    "day": "Lunes",
    "id": 1,
    "place": "Centro de Aprendizaje Global",
    "teacher_id": 1,
    "time": "13:00",
    "url": null,
    "workshop_id": 1,
  },
  "option_id": 1,
  "student": {
    "applicant": {
      "apellido_materno": "Valadez",
      "apellido_paterno": "Aguilera",
      "carrera": "Agrobiotecnologia (AGB)",
      "ciclo_ingreso": "2022B",
      "codigo": "0987654321",
      "email": "juanga@elnoanoa.mx",
      "externo": false,
      "genero": "M",
      "id": 2,
      "institutional_email": "alberto.aguilera@alumnos.udg.mx",
      "nombre": "Alberto",
      "telefono": "0001234567",
    },
    "applicantId": 2,
    "ciclo_actual": "2022B",
    "codigo": "0987654321",
    "curso": "en",
    "desertor": false,
    "groupId": 2,
    "groupObject": {
      "aula": "F4",
      "ciclo": "2022B",
      "course": "en",
      "id": 2,
      "name": "E4-1",
      "nivel": 4,
      "registrados": 30,
      "teacherId": 1,
      "time": "08:00-09:00",
    },
    "id": 2,
    "nivel": 4,
    "situacion": null,
  },
  "student_id": 2,
}
`;

exports[`gets a teacher option with all reservations for attendance 2`] = `
{
  "attended": false,
  "create_time": Any<Date>,
  "id": 2,
  "option": {
    "active": true,
    "day": "Lunes",
    "id": 1,
    "place": "Centro de Aprendizaje Global",
    "teacher_id": 1,
    "time": "13:00",
    "url": null,
    "workshop_id": 1,
  },
  "option_id": 1,
  "student": {
    "applicant": {
      "apellido_materno": "Ocasio",
      "apellido_paterno": "Martinez",
      "carrera": "Abogado (DECH)",
      "ciclo_ingreso": "2022A",
      "codigo": "1234567890",
      "email": "bad@bunny.pr",
      "externo": false,
      "genero": "M",
      "id": 1,
      "institutional_email": "benito.martinez@cusur.udg.mx",
      "nombre": "Benito Antonio",
      "telefono": "0001234567",
    },
    "applicantId": 1,
    "ciclo_actual": "2022B",
    "codigo": "1234567890",
    "curso": "en",
    "desertor": false,
    "groupId": 2,
    "groupObject": {
      "aula": "F4",
      "ciclo": "2022B",
      "course": "en",
      "id": 2,
      "name": "E4-1",
      "nivel": 4,
      "registrados": 30,
      "teacherId": 1,
      "time": "08:00-09:00",
    },
    "id": 3,
    "nivel": 4,
    "situacion": null,
  },
  "student_id": 3,
}
`;

exports[`makes a reservation 1`] = `
{
  "attended": false,
  "create_time": 2022-09-23T00:00:00.000Z,
  "id": "1",
  "option": {
    "active": true,
    "day": "Lunes",
    "id": 1,
    "place": "Centro de Aprendizaje Global",
    "teacher": {
      "id": 1,
      "nombre": "Gonzalo Rocha",
    },
    "teacher_id": 1,
    "time": "13:00",
    "url": null,
    "workshop": {
      "description": "Conversation Club",
      "id": 1,
      "levels": [
        1,
        2,
        3,
        4,
        5,
        6,
      ],
      "name": "Conversation",
    },
    "workshop_id": 1,
  },
  "option_id": 1,
  "student": {
    "applicantId": 1,
    "ciclo_actual": "2022B",
    "codigo": "1234567890",
    "curso": "en",
    "desertor": false,
    "groupId": 2,
    "id": 3,
    "nivel": 4,
    "situacion": null,
  },
  "student_id": 3,
}
`;
